guides/recipes/anti-patterns.md

```diff
@@ -0,0 +1,843 @@
+  end
+end
+```
+
+### Additional Remarks
+
+Design your queue structure based on these job characteristics:
+
+| Characteristic | Considerations |
+|----------------|---------------|
+| **Duration** | Separate quick jobs from long-running ones |
+| **Resource usage** | Limit concurrency for resource-intensive operations |
+| **Priority** | Use both queues and priority values for fine-grained control |
+| **External dependencies** | Create dedicated queues for rate-limited external services |
+| **Business importance** | Ensure critical operations get processed first |
+
+The Smart Engine will dynamically adjust concurrency based on job execution patterns, but it works best when you start with thoughtfully designed queue structure.
+
+## Anti-Pattern 7: Insufficient Error Handling in Jobs
+
+### Problem
+
+Jobs with inadequate error handling lead to:
+
+- Unnecessary retries that waste resources
+- Masking of issues that should trigger alerts
+- Lack of context for troubleshooting
+- Unpredictable system behavior under failure conditions
+
+### Example
+
+```elixir
+defmodule MyApp.Workers.UserSynchronizer do
+  use Oban.Worker, queue: :sync, max_attempts: 20
+
+  @impl Oban.Worker
+  def perform(%{args: %{"user_id" => user_id}}) do
+    # Poor error handling - will retry even for expected errors
+    case MyApp.Users.fetch(user_id) do
+      {:ok, user} ->
+        MyApp.ExternalService.sync_user(user)
+
+      {:error, _reason} ->
+        # Generic error, will trigger retry without context
+        raise "Failed to sync user #{user_id}"
+    end
+  end
+end
+```
+
+### Refactoring
+
+Implement thoughtful error handling with appropriate retry behavior:
+
+```elixir
+defmodule MyApp.Workers.UserSynchronizer do
+  use Oban.Worker, queue: :sync, max_attempts: 5
+  require Logger
+
+  @impl Oban.Worker
+  def perform(%{args: %{"user_id" => user_id} = args, attempt: attempt}) do
+    case MyApp.Users.fetch(user_id) do
+      {:ok, user} ->
+        case MyApp.ExternalService.sync_user(user) do
+          {:ok, _} ->
+            :ok
+
+          {:error, %{status: 400, reason: reason}} ->
+            # Client error - don't retry
+            Logger.warn("User sync failed with client error: #{reason}", args: args)
+            {:discard, :client_error}
+
+          {:error, %{status: status}} when status >= 500 ->
+            # Server error - worth retrying with backoff
+            backoff = :math.pow(2, attempt) |> round()
+            Logger.warn("User sync failed with server error: #{status}", args: args)
+            {:snooze, backoff}
+
+          {:error, :timeout} ->
+            # Network issue - worth retrying
+            Logger.warn("User sync timed out", args: args)
+            {:error, :timeout}
+        end
+
+      {:error, :not_found} ->
+        # Don't retry when the user doesn't exist
+        Logger.info("Cannot sync user #{user_id} - not found", args: args)
+        {:discard, :user_not_found}
```

Comment:
Maybe a good example of when to use `cancel`? As the job became a no-op

Reply:
Definitely use `{:cancel, :user_not_found}` rather than `{:discard, _}`. And don't include logging.